Cannot create multiple closures of an xfunction expression

If I understand correctly, to use a passed-in expression multiple times in a function, I need to create a closure of the expression for each use. Like this:

```c++
template<typename E>
auto normalize(E&& expr)
{
    return E{expr} / xt::view(xt::sum(E{expr}, {1}), xt::all(), xt::newaxis());
}
```

The code above works with containers and views but not with arithmetic operations:

```c++
xt::xtensor<double, 2> freqs({10, 5}, 1);
xt::xtensor<double, 2> other_freqs({10, 5}, 2);
xt::xtensor<double, 2> probs({10, 5});
probs = normalize(freqs); // OK
probs = normalize(xt::view(freqs, xt::all(), xt::all())); // OK
probs = normalize(freqs + other_freqs); // Error!
```

The error message is shown below. It looks like the copy constructor of `xfunction` (which I expect to be invoked here) is not selected.

```
/usr/local/include/xtensor/xfunction.hpp:543:41: error: no matching constructor for initialization
      of 'functor_type' (aka 'xt::detail::plus<double>')
        : m_e(std::forward<CTA>(e)...), m_f(std::forward<Func>(f)), m_shape(xtl::make_seque...
                                        ^   ~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/xtensor/xoperation.hpp:108:33: note: candidate constructor
      (the implicit copy constructor) not viable: no known conversion from
      'xt::xfunction<xt::detail::plus<double>, double, const
      xt::xtensor_container<xt::uvector<double, std::__1::allocator<double> >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &, const
      xt::xtensor_container<xt::uvector<double, std::__1::allocator<double> >, 2,
      xt::layout_type::row_major, xt::xtensor_expression_tag> &>' to
      'const xt::detail::plus<double>' for 1st argument
        BINARY_OPERATOR_FUNCTOR(plus, +);
                                ^
/usr/local/include/xtensor/xoperation.hpp:100:72: note: expanded from macro
      'BINARY_OPERATOR_FUNCTOR'
...
```

--- a/include/xtensor/xfunction.hpp
+++ b/include/xtensor/xfunction.hpp
@@ -286,7 +286,7 @@ namespace xt
 
     protected:
 
-        template <class Func, class... CTA, class U = std::enable_if_t<!std::is_base_of<Func, self_type>::value>>
+        template <class Func, class... CTA, class U = std::enable_if_t<!std::is_base_of<std::decay_t<Func>, self_type>::value>>
         xfunction_base(Func&& f, CTA&&... e) noexcept;
 
         ~xfunction_base() = default;
@@ -511,7 +511,7 @@ namespace xt
         using self_type = xfunction<F, R, CT...>;
         using base_type = xfunction_base<F, R, CT...>;
 
-        template <class Func, class... CTA, class U = std::enable_if_t<!std::is_base_of<Func, self_type>::value>>
+        template <class Func, class... CTA, class U = std::enable_if_t<!std::is_base_of<std::decay_t<Func>, self_type>::value>>
         xfunction(Func&& f, CTA&&... e) noexcept;
 
         ~xfunction() = default;

--- a/include/xtensor/xoptional.hpp
+++ b/include/xtensor/xoptional.hpp
@@ -361,7 +361,7 @@ namespace xt
         using value_functor = typename F::template rebind<typename R::value_type>::type;
         using flag_functor = detail::optional_bitwise<bool>;
 
-        template <class Func, class U = std::enable_if<!std::is_base_of<Func, self_type>::value>>
+        template <class Func, class U = std::enable_if<!std::is_base_of<std::decay_t<Func>, self_type>::value>>
         xoptional_function(Func&& func, CT... e) noexcept;
 
         ~xoptional_function() = default;

--- a/test/test_xfunction.cpp
+++ b/test/test_xfunction.cpp
@@ -52,6 +52,20 @@ namespace xt
         }
     }
 
+    template<typename E>
+    auto normalize(E&& expr)
+    {
+        return E{ expr } / xt::view(xt::sum(E{ expr }, { 1 }), xt::all(), xt::newaxis());
+    }
+
+    TEST(xfunction, copy_constructor)
+    {
+        xt::xarray<double> freqs({ 10, 5 }, 1);
+        xt::xarray<double> other_freqs({ 10, 5 }, 2);
+        // Compilation test
+        xarray<double> probs = normalize(freqs + other_freqs);
+    }
+
     TEST(xfunction, broadcast_shape)
     {
         using shape_type = layout_result<>::shape_type;

